File: cpp/UCLA/libUCLA/USender.cpp
#include "USender.h"
#include "config.h"

#ifdef UCLA_HAVE_UNIX
#include <xs/xs.hpp>
#else
#include <xs.hpp>
#endif

using namespace UCLA;

USender::USender(UConfig &config, bool autostart){
	this->_isStarted = false;
	this->_ctx = NULL;
	this->_sock = NULL;
	this->_endpoint = NULL;

	CopyEndpoint(config.Endpoint());

	if(autostart){
		Start();
	}
}

USender::USender(USender&& that){
	if(this->_isStarted)
	{
		this->CleanUpXS();
	}

	this->_isStarted = false;
	this->_ctx = NULL;
	this->_sock = NULL;
	this->_endpoint = NULL;

	CopyEndpoint(that.Endpoint());

	that.~USender();
}

USender& USender::operator=(USender&& that){
	if (this != &that){
		if(this->_isStarted)
		{
			this->CleanUpXS();
		}

		this->_isStarted = false;
		this->_ctx = NULL;
		this->_sock = NULL;

		CopyEndpoint(that.Endpoint());

		that.~USender();
	}

	return *this;
}

void USender::CopyEndpoint(const char* endpoint){
	using namespace Utils;

	if(this->_endpoint != NULL)
		delete[] this->_endpoint;

	int strLength = strlen(endpoint) + 1;

	char* tmp_endpoint = new char[strLength];
	this->_endpoint = tmp_endpoint;

	m_strlcpy(this->_endpoint, strLength, endpoint);
}

void USender::Start(void){
	if(this->_isStarted){
		return;
	}

	try{
		this->_ctx = new xs::context_t();
		this->_sock = new xs::socket_t(*(static_cast<xs::context_t*>(this->_ctx)), XS_PUSH);

		static_cast<xs::socket_t*>(this->_sock)->connect(this->_endpoint);
	}catch(xs::error_t &err){
		throw UException(err.what());
	}

	this->_isStarted = true;
}

void USender::SendData(const char *buf, size_t len, bool nonBlocking){
	if(!this->_isStarted){
		this->Start();
	}

	try{
		static_cast<xs::socket_t*>(this->_sock)->send(buf, len, nonBlocking? XS_DONTWAIT : 0);
	}catch(xs::error_t &err){
		throw UException(err.what());
	}
}

bool USender::IsStarted(void) const{
	return this->_isStarted;
}

void USender::CleanUpXS(void){
	if(this->_sock != NULL){
		static_cast<xs::socket_t*>(this->_sock)->close();
		delete static_cast<xs::socket_t*>(this->_sock);

		this->_sock = NULL;
	}

	if(this->_ctx != NULL){
		delete static_cast<xs::context_t*>(this->_ctx);

		this->_ctx = NULL;
	}
}

USender::~USender(void){
	this->_isStarted = false;

	delete[] this->_endpoint;
	this->_endpoint = NULL;

	this->CleanUpXS();
}

File: cs/UCLA/libUCLA/USender.cs
using CrossroadsIO;
using System;

namespace libUCLA {
    public class USender : IDisposable {

        /// <summary>
        /// True if connection has been started.
        /// </summary>
        public bool IsStarted {
            get { return this.isStarted; }
        }

        private bool isStarted = false;
        private Context ctx;
        private Socket sock;
        private string endpoint;

        private bool _disposed;

        /// <summary>
        /// UCLA Sender constructor.
        /// Creates sender instance from specified config.
        /// </summary>
        /// <param name="config">Reference to configuration object.</param>
        /// <param name="autostart">Tells if connection will be started automatically. 
        /// Otherwise you will have to call Start method. Default value = false.</param>
        public USender(UConfig config, bool autostart = false) {
            this.endpoint = config.Endpoint;

            if (autostart) {
                this.Start();
            }
        }

        /// <summary>
        /// Starts connection to a receiver.
        /// </summary>
        public void Start() {
            if (this.isStarted) return;

            try {
                this.ctx = Context.Create();
                this.sock = this.ctx.CreateSocket(SocketType.PUSH);

                this.sock.Connect(this.endpoint);
            }
            catch (XsException ex) {
                throw new UException(ex.Message, ex);
            }

            this.isStarted = true;
        }

        /// <summary>
        /// Sends data to a receiver in (hopefully) blocking mode.    
        /// </summary>
        /// <param name="data">Buffer with data.</param>
        public void SendData(byte[] data) {
            if (!this.isStarted) {
                this.Start();
            }

            try {
                this.sock.Send(data, data.Length, SocketFlags.None);
            }
            catch (XsException ex) {
                throw new UException(ex.Message, ex);
            }
        }

        public void Dispose() {
            this.Dispose(true);
            GC.SuppressFinalize(this);
        }

        protected virtual void Dispose(bool disposing) {
            if (!this._disposed) {
                if (disposing) {
                    try {
                        this.sock.Close();
                        this.ctx.Terminate();

                        this.sock.Dispose();
                        this.ctx.Dispose();
                    }
                    catch (XsException ex) {
                        throw new UException(ex.Message, ex);
                    }
                }

                this._disposed = true;
            }
        }

        private void EnsureNotDisposed() {
            if (this._disposed) {
                throw new ObjectDisposedException(GetType().FullName);
            }
        }
    }
}
